src/sync/Synchroniser.js

```diff
@@ -238,65 +249,45 @@ export class Synchroniser {
    * @return {Promise} Resolves if successful, or passes up any error thrown
    */
   pull = async () => {
-    this.setProgressMessage('Pulling changes from the server');
-    this.setProgress(0);
-    this.setTotal(0);
-    const serverURL = this.settings.get(SYNC_URL);
-    const thisSiteId = this.settings.get(SYNC_SITE_ID);
-    const serverId = this.settings.get(SYNC_SERVER_ID);
-    await this.recursivePull(serverURL, thisSiteId, serverId, 0);
-  };
+    let waitingRecordCount = await this.getWaitingRecordCount();
+    let total = waitingRecordCount;
+    let progress = 0;
 
-  /**
-   * Recursively checks how many records left to pull, pulls in a batch, and calls
-   * itself
-   * @param  {string} serverURL  The URL of the sync server
-   * @param  {string} thisSiteId The sync ID of this sync site
-   * @param  {string} serverId   The sync ID of the server
-   * @return {Promise}          Resolves if successful, or passes up any error thrown
-   */
-  recursivePull = async (serverURL, thisSiteId, serverId, currentTotal) => {
-    const authHeader = this.authenticator.getAuthHeader();
-    const waitingRecordCount = await this.getWaitingRecordCount(
-      serverURL,
-      thisSiteId,
-      serverId,
-      authHeader,
-    );
-    // Allow total to increase in case more records are added to the server sync queue during sync
-    const newTotal = Math.max(waitingRecordCount, currentTotal);
-    this.setTotal(newTotal);
-    if (waitingRecordCount === 0) return; // Done recursing through records
+    this.setProgressMessage('Pulling changes from the server');
+    this.setProgress(progress);
+    this.setTotal(total);
 
-    // Get a batch of records and integrate them
-    const incomingRecords = await this.getIncomingRecords(
-      serverURL,
-      thisSiteId,
-      serverId,
-      authHeader,
-      BATCH_SIZE,
-    );
-    this.integrateRecords(incomingRecords);
-    await this.acknowledgeRecords(serverURL, thisSiteId, serverId, authHeader, incomingRecords);
-    this.incrementProgress(incomingRecords.length);
+    // Pull this.batchSize amount of records at a time from server
+    while (progress < total) {
+      // Get a batch of records and integrate them
+      const incomingRecords = await this.getIncomingRecords();
+      this.integrateRecords(incomingRecords);
+      await this.acknowledgeRecords(incomingRecords);
+      this.incrementProgress(incomingRecords.length);
+      progress += incomingRecords.length;
 
-    // Recurse to get the next batch of records from the server
-    await this.recursivePull(serverURL, thisSiteId, serverId, newTotal);
+      // Check no new records added to incoming queue on last iteration.
+      if (progress >= total) {
+        waitingRecordCount = await this.getWaitingRecordCount();
+        if (waitingRecordCount > 0) {
+          total += waitingRecordCount;
```

Comment:
I think this should be `total = progress + waitingRecordCount`, otherwise you can get progress exceeding total by maximums of INITIAL_BATCH_SIZE

Reply:
Otherwise the PR looks very good

Reply:
I'm not sure what you mean, as I understand it:
Currently:
1. 19/20
2. 39/20 (very briefly)
3. 39/60 (there was another 40 records on server)
4. 59/60
5. 60/60 (incomingRecords.length was 1)

Your suggestion:
1. 19/20
2. 39/20
3. 39/79 (there were another 40 records on server)
4. 59/79
5. 60/79 (incomingRecords.length was 1)
6. Falsely claiming an extra 19 records waiting, server never returns them, we end up in infinite loop? 

Step 2 can be improved by checking and updating total before calling incrementProgress, this way in the UI total is updated before progress.

Reply:
Correction (I think):
Currently:
1. 19/20
2. 39/20 (very briefly)
3. 39/41 (there was another ~~40~~ 21 records on server, as we've retrieved 20 of the total 60 since the last query on waiting)
4. 59/41 (
5. 59/60 (incomingRecords.length was 1)

Your suggestion:
1. 19/20
2. 39/20
3. 39/60 (there were another ~~40~~ 21 records on server)
4. 59/60
5. 60/60 (incomingRecords.length was 1)
6. ~~Falsely claiming an extra 19 records waiting, server never returns them, we end up in infinite loop?~~

Reply:
Maybe should also check for incomingRecords.length = 0 and exit loop, this is before integrating records

Reply:
@Chris-Petty that requested change is not really necessary I was just being pedantic.
But I do think we need a check to make sure the loop is not infinite (if server for some reason has less records then the total)